File: ocaml/basics/03_loops_control_flow.ml
(* TOPIC: Loops & Control Flow | ocaml 03_loops_control_flow.ml *)

(* TODO 1: Imperative for loop — only ascending, no return value *)
(*   for i = 1 to 10 do
       Printf.printf "%d " i
     done;
     for i = 10 downto 1 do  (* downto for descending *)
       Printf.printf "%d " i
     done  *)

(* TODO 2: while loop — mutable state required *)
(*   let i = ref 0 in
     while !i < 10 do
       Printf.printf "%d " !i;
       i := !i + 1
     done  *)

(* TODO 3: Functional recursion — preferred over loops in OCaml *)
(*   let rec range a b =
       if a > b then [] else a :: range (a+1) b
     let rec iterate f n x =
       if n = 0 then x else iterate f (n-1) (f x)  *)

(* TODO 4: Pattern matching — exhaustive, no fallthrough *)
(*   let describe n = match n with
       | 0 -> "zero"
       | 1 -> "one"
       | n when n < 0 -> "negative"
       | n when n mod 2 = 0 -> "positive even"
       | _ -> "positive odd"
     (* Match with tuples: *)
     let sign_pair (a, b) = match (a > 0, b > 0) with
       | (true, true) -> "both positive"
       | (false, false) -> "both negative"
       | _ -> "mixed"  *)

(* TODO 5: if expressions — always returns a value *)
(*   let abs x = if x >= 0 then x else -x
     let max_val a b = if a > b then a else b
     (* Note: if without else returns unit, so branches must match types *)  *)

(* TODO 6: List.iter and List.iteri — side-effect loops over lists *)
(*   List.iter (Printf.printf "%d ") [1;2;3;4;5];
     List.iteri (fun i x -> Printf.printf "[%d]=%d " i x) [10;20;30]  *)

(* TODO 7: Sequence operations with |> (pipe operator) *)
(*   [1;2;3;4;5;6;7;8;9;10]
     |> List.filter (fun x -> x mod 2 = 0)
     |> List.map (fun x -> x * x)
     |> List.fold_left (+) 0
     |> Printf.printf "Sum of even squares: %d\n"  *)

(* TODO 8: Recursion with accumulator — FizzBuzz *)
(*   let fizzbuzz n =
       let rec go i =
         if i > n then ()
         else begin
           (match (i mod 3, i mod 5) with
             | (0, 0) -> print_endline "FizzBuzz"
             | (0, _) -> print_endline "Fizz"
             | (_, 0) -> print_endline "Buzz"
             | _      -> print_int i; print_newline ());
           go (i + 1)
         end
       in go 1  *)

let () = print_endline "TODO: implement control flow exercises"

File: kotlin/basics/03_loops_control_flow.kt
package basics
// TOPIC: Loops & Control Flow | kotlinc 03_loops_control_flow.kt -include-runtime -d out.jar && java -jar out.jar
// Docs: https://kotlinlang.org/docs/control-flow.html

fun main() {
    // TODO 1: Ranges — for (i in 1..10), downTo, step, until
    //   - Print 1..10 inclusive
    //   - Print 10 downTo 1
    //   - Print 0 until 10 step 2 (evens, exclusive upper bound)
    //   - Use a range in an if condition: if (score in 90..100)

    // TODO 2: when expression — replaces switch, use as expression returning value
    //   - val grade = when (score) { in 90..100 -> "A"; in 80..89 -> "B"; else -> "F" }
    //   - when on a sealed class (exhaustive, no else needed)
    //   - when without argument: when { x > 0 -> "positive"; else -> "non-positive" }

    // TODO 3: while & do-while — FizzBuzz 1-100
    //   - Implement FizzBuzz using a while loop
    //   - Show do-while: var input: String; do { input = readLine()!! } while (input.isBlank())

    // TODO 4: break/continue with labels — @outer label to break nested loop
    //   - Nested loop searching for a value, break@outer when found
    //   - continue@outer to skip to next outer iteration
    //   - Show the difference between plain break and labeled break

    // TODO 5: Collection operations — filter, map, reduce, fold, groupBy, partition
    //   - val nums = (1..20).toList()
    //   - filter { it % 2 == 0 }, map { it * it }
    //   - reduce { acc, n -> acc + n } vs fold(0) { acc, n -> acc + n }
    //   - groupBy { if (it % 3 == 0) "fizz" else "other" }
    //   - partition { it > 10 } — returns Pair<List,List>

    // TODO 6: Sequences — asSequence() for lazy evaluation, compare with eager List ops
    //   - (1..1_000_000).asSequence().filter { it % 2 == 0 }.map { it * it }.first()
    //   - Compare with eager list version: note sequences short-circuit on terminal op
    //   - generateSequence(1) { it + 1 }.take(10).toList()

    // TODO 7: forEach vs for — when to use each, forEach with index (forEachIndexed)
    //   - for loop: prefer when you need break/continue
    //   - forEach: cleaner for full iteration, no early exit
    //   - forEachIndexed { index, value -> println("$index: $value") }
    //   - mapIndexed { index, value -> "$index=$value" }

    // TODO 8: repeat — repeat(5) { println("Hello $it") }
    //   - repeat(5) { println("Hello $it") } — it is the zero-based iteration index
    //   - Use repeat to build a string: val s = buildString { repeat(3) { append("ha") } }
    //   - Show difference vs for loop — repeat is for simple side-effect repetition
}
